src/bin/givc-agent.rs

```diff
@@ -0,0 +1,138 @@
+use clap::Parser;
+use givc::admin::client::AdminClient;
+use givc::endpoint::{EndpointConfig, TlsConfig};
+use givc::pb;
+use givc::systemd_api::server::SystemdService;
+use givc::types::*;
+use givc::utils::naming::*;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use tonic::transport::Server;
+
+#[derive(Debug, Parser)] // requires `derive` feature
+#[command(name = "givc-agent")]
+#[command(about = "A givc agent", long_about = None)]
+struct Cli {
+    #[arg(long, env = "NAME")]
+    name: String,
+
+    #[arg(long, env = "ADDR", default_missing_value = "127.0.0.1")]
+    addr: String,
+    #[arg(long, env = "PORT", default_missing_value = "9001")]
+    port: u16,
+
+    #[arg(long, env = "TLS", default_missing_value = "false")]
+    use_tls: bool,
+
+    #[arg(long, env = "TYPE")]
+    r#type: u32,
+
+    #[arg(long, env = "SUBTYPE", default_missing_value = "14")]
+    subtype: u32,
+
+    #[arg(long, env = "CA_CERT")]
+    ca_cert: Option<PathBuf>,
+
+    #[arg(long, env = "HOST_CERT")]
+    host_cert: Option<PathBuf>,
+
+    #[arg(long, env = "HOST_KEY")]
+    host_key: Option<PathBuf>,
+
+    #[arg(long, env = "ADMIN_SERVER_ADDR", default_missing_value = "127.0.0.1")]
+    admin_server_addr: String,
+    #[arg(long, env = "ADMIN_SERVER_PORT", default_missing_value = "9000")]
+    admin_server_port: u16,
+
+    #[arg(
+        long,
+        env = "SERVICES",
+        use_value_delimiter = true,
+        value_delimiter = ','
+    )]
+    services: Option<Vec<String>>,
+}
+
+// FIXME: should be in src/lib.rs: mod pb {}, but doesn't work
+mod kludge {
+    pub const FILE_DESCRIPTOR_SET: &[u8] =
+        tonic::include_file_descriptor_set!("systemd_descriptor");
+}
+
+#[tokio::main]
+async fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
+    let cli = Cli::parse();
+    println!("CLI is {:#?}", cli);
+
+    let addr = SocketAddr::new(cli.addr.clone().parse().unwrap(), cli.port);
+
+    let agent_service_name = format_service_name(&cli.name);
+
+    let mut builder = Server::builder();
+
+    let tls = if cli.use_tls {
+        let tls = TlsConfig {
+            ca_cert_file_path: cli.ca_cert.ok_or(String::from("required"))?,
+            cert_file_path: cli.host_cert.ok_or(String::from("required"))?,
```

Comment:
No need to create a `String`, an `&str` works too; the text should probably have a bit of context of what is required.